README.md
## ms-jpa-utility
this project is a shared common utility project for microservice projects as a pom dependency to reduce boilerplate code in multiple projects.

### Project Descriptions:
build with java-doc:
- if your windows account is for example "myuser" and you have "C:\Users\MyUser\.m2\repository\com\motaharinia\ms-jpa-utility" folder, remove it.
- open ms-jpa-utility project by IntellliJ IDEA(enable auto import , Windows defender automatic fix) and check in "project structure" that jdk 11 is selected.
- in IntellliJ IDEA open right side maven panel and select lifecycle/install and run maven to build it in "C:\Users\MyUser\.m2\repository\com\motaharinia\ms-jpa-utility"

### Change Log:
- 0.0.1 : initial project
- 0.0.2 : add type conversion for Instant,LocalDate,LocalDateTime in CustomMapper class. add languages for customvalidation messages

### IntellliJ IDEA Configurations:
- IntelijIDEA: Help -> Edit Custom Vm Options -> add these two line:
  - -Dfile.encoding=UTF-8
  - -Dconsole.encoding=UTF-8
- IntelijIDEA: File -> Settings -> Editor -> File Encodings-> Project Encoding: form "System default" to UTF-8. Maybe it affected somehow.
- IntelijIDEA: File -> Settings -> Editor -> General -> Code Completion -> check "show the documentation popup in 500 ms"
- IntelijIDEA: File -> Settings -> Editor -> General -> Auto Import -> check "Optimize imports on the fly (for current project)"
- IntelijIDEA: File -> Settings -> Editor -> Color Scheme -> Color Scheme Font -> Scheme: Default -> uncheck "Show only monospaced fonts" and set font to "Tahoma"
- IntelijIDEA: File -> Settings -> Editor -> Proofreading -> add "farsi" language for graze spell check
- IntelijIDEA: Run -> Edit Configuration -> Spring Boot -> XXXApplication -> Configuration -> Environment -> VM Options: -Dspring.profiles.active=dev
- IntelijIDEA: Run -> Edit Configuration -> Spring Boot -> XXXApplication -> Code Coverage -> Fix the package in include box

<hr/>
<a href="mailto:eng.motahari@gmail.com?"><img src="https://img.shields.io/badge/gmail-%23DD0031.svg?&style=for-the-badge&logo=gmail&logoColor=white"/></a>




CHANGELOG.md
# Change Log
#### [0.0.1] - 2017-03-15
* Added:
* Changed:
* Fixed:

#### [0.0.2] - 2021-02-04
* Added: 
  * DbColumnConfiguration.COLUMN_DEFINITION_TIME
  * DbColumnConfiguration.COLUMN_DEFINITION_TIMESTAMP
  * DbColumnConfiguration.COLUMN_DEFINITION_TIME_WITH_TIMEZONE
  * DbColumnConfiguration.COLUMN_DEFINITION_TIMESTAMP_WITH_TIMEZONE
* Changed:
  * MsUtility version from 0.0.1 to 0.0.2
  * GenericEntity to CustomEntity with LocalDateTime fields
* Fixed:

#### [0.0.3] - 2021-02-07
* Added:
* Changed:
  * update MsUtility version to 0.0.3
* Fixed:

#### [0.0.4] - 2021-02-09
* Added:
* Changed:
  * update MsUtility version to 0.0.4
* Fixed:

#### [0.0.5] - 2021-02-11
* Added:
* Changed:
  * makes the code more clean
  * update MsUtility version to 0.0.5
* Fixed:

#### [0.0.6] - 2021-02-22
* Added:
* Changed:
  * update MsUtility version to 0.0.6
* Fixed:

#### [0.0.7] - 2021-03-12
* Added:
* Changed:
  * update MsUtility version to 0.0.7
* Fixed:

#### [0.0.8] - 2021-03-31
* Added:
* Changed:
  * update MsUtility version to 0.0.8
* Fixed:

#### [0.0.9] - 2021-03-31
* Added:
* Changed:
  * update MsUtility version to 0.0.9
* Fixed:

#### [0.0.10] - 2021-04-06
* Added:
* Changed:
  * update MsUtility version to 0.0.10
* Fixed:

#### [0.0.11] - 2021-04-23
* Added:
* Changed:
  * update MsUtility version to 0.0.11
* Fixed:

#### [0.0.12] - 2021-05-10
* Added:
* Changed:
  * update MsUtility version to 0.0.12
* Fixed:

#### [0.0.13] - 2021-05-13
* Added:
* Changed:
  * update MsUtility version to 0.0.13
  * moved search/filter and search/data to MsUtility version to 0.0.13
  * deleted unused application.properties, application-dev.properties, application-com.properties
* Fixed:

#### [0.0.14] - 2021-05-15
* Added:
* Changed:
  * update MsUtility version to 0.0.14
* Fixed:

#### [0.0.15] - 2021-05-22
* Added:
* Changed:
  * update MsUtility version to 0.0.15
* Fixed:

#### [0.0.16] - 2021-06-09
* Added:
* Changed:
  * update MsUtility version to 0.0.16
* Fixed:

#### [0.0.17] - 2021-06-13
* Added:
* Changed:
  * update MsUtility version to 0.0.17
* Fixed:

#### [0.0.18] - 2021-06-16
* Added:
* Changed:
  * update MsUtility version to 0.0.18
* Fixed:

#### [0.0.19] - 2021-06-19
* Added:
* Changed:
  * update MsUtility version to 0.0.19
* Fixed:

#### [0.0.20] - 2021-06-19
* Added:
* Changed:
  * update MsUtility version to 0.0.20
* Fixed:

#### [0.0.21] - 2021-06-19
* Added:
* Changed:
  * update MsUtility version to 0.0.21
* Fixed:

#### [0.0.22] - 2021-07-08
* Added:
* Changed:
  * update MsUtility version to 0.0.22
* Fixed:

#### [0.0.23] - 2021-07-09
* Added:
* Changed:
  * update MsUtility version to 0.0.23
* Fixed:

#### [0.0.24] - 2021-07-19
* Added:
* Changed:
  * update MsUtility version to 0.0.24
  * changed artifactId to ms-jpa-utility
* Fixed:

#### [0.0.25] - 2021-07-31
* Added:
* Changed:
  * update ms-utility version to 0.0.25
* Fixed:

#### [0.0.26] - 2021-08-04
* Added:
* Changed:
  * update ms-utility version to 0.0.26
* Fixed:

#### [0.0.27] - 2021-08-06
* Added:
* Changed:
  * update ms-utility version to 0.0.27
* Fixed:

#### [0.0.28] - 2021-08-22
* Added:
  * added CustomIntegerSetAttributeConverter, CustomLongListAttributeConverter, CustomLongSetAttributeConverter, CustomStringSetAttributeConverter
* Changed:
  * changed CustomIntegerAttributeConverter name to CustomIntegerListAttributeConverter
  * changed CustomStringAttributeConverter name to CustomStringListAttributeConverter
  * update ms-utility version to 0.0.28
* Fixed:

#### [0.0.29] - 2021-09-01
* Added:
* Changed:
  * update ms-utility version to 0.0.29
* Fixed:

#### [0.0.30] - 2021-09-21
* Added:
* Changed:
  * update ms-utility version to 0.0.30
* Fixed:

#### [0.0.31] - 2021-10-02
* Added:
* Changed:
  * update ms-utility version to 0.0.31
* Fixed:

#### [0.0.32] - 2021-10-06
* Added:
* Changed:
  * update ms-utility version to 0.0.32
* Fixed:

#### [0.0.33] - 2021-10-16
* Added:
* Changed:
  * update ms-utility version to 0.0.33
* Fixed:

#### [0.0.34] - 2021-10-17
* Added:
* Changed:
  * update ms-utility version to 0.0.34
* Fixed:

#### [0.0.35] - 2021-10-19
* Added:
* Changed:
  * update ms-utility version to 0.0.35
* Fixed:

#### [0.0.36] - 2021-10-20
* Added:
  * added CustomPageImpl to use in test methods and anywhere we want read Page return in RestTemplate
* Changed:
  * update ms-utility version to 0.0.36
* Fixed:

#### [0.0.37] - 2021-11-06
* Added:
* Changed:
  * update ms-utility version to 0.0.37
* Fixed:

#### [0.0.38] - 2021-11-10
* Added:
* Changed:
  * update ms-utility version to 0.0.38
* Fixed:

#### [0.0.39] - 2021-11-14
* Added:
* Changed:
  * update ms-utility version to 0.0.39
* Fixed:

#### [0.0.40] - 2021-11-16
* Added:
  * added CustomPageResponseDto to return Page as DTO
  * added CustomDocument to use in MongoDB documents
  * added CustomCounter to use in MongoDB services
* Changed:
  * update ms-utility version to 0.0.40
* Fixed:

#### [0.0.41] - 2021-11-21
* Added:
* Changed:
  * update ms-utility version to 0.0.41
* Fixed:

#### [0.0.42] - 2021-12-05
* Added:
* Changed:
  * update ms-utility version to 0.0.42
* Fixed:

#### [0.0.43] - 2021-12-07
* Added:
* Changed:
  * update ms-utility version to 0.0.43
* Fixed:

#### [0.0.44] - 2021-12-08
* Added:
* Changed:
  * update ms-utility version to 0.0.44
* Fixed:

#### [0.0.45] - 2021-01-22
* Added:
* Changed:
    * update ms-utility version to 0.0.45
* Fixed:

src/main/java/com/motaharinia/msjpautility/entity/DbColumnDefinition.java
package com.motaharinia.msjpautility.entity;

import org.jetbrains.annotations.NotNull;
import org.springframework.util.ObjectUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

/**
* @author https://github.com/motaharinia<br>
 * این کلاس برای مدیریت یکپارچه متغیرهای مورد نیاز برای دیتابیس در سطح انتیتی ها آماده شده است
 */
public interface DbColumnDefinition {

    /**
     * تعریف میزان محاسه عدد اعشار تا 5 رقم
     */
    Integer BIG_DECIMAL_SCALE = 5;
    /**
     * تعریف نوع داده میزان پول با طول عدد صحیح 19 رقم و طول اعشار 4 رقم
     */
    String COLUMN_DEFINITION_AMOUNT_194 = "DECIMAL(19,4)";
    /**
     * تعریف نوع داده اعشاری با طول عدد صحیح 8 رقم و طول اعشار 4 رقم
     */
    String COLUMN_DEFINITION_DOUBLE_84 = "DECIMAL(8,4)";
    /**
     * تعریف نوع داده اعشاری با طول عدد صحیح 14 رقم و طول اعشار 4 رقم
     */
    String COLUMN_DEFINITION_DOUBLE_144 = "DECIMAL(14,4)";
    /**
     * تعریف نوع داده تاریخ معادل java.time.LocalDate
     */
    String COLUMN_DEFINITION_DATE = "DATE";
    /**
     * تعریف نوع داده زمان معادل java.time.LocalTime
     */
    String COLUMN_DEFINITION_TIME = "TIME";
    /**
     * تعریف نوع داده تاریخ-زمان معادل java.time.LocalDateTime
     */
    String COLUMN_DEFINITION_TIMESTAMP = "TIMESTAMP";
    /**
     * تعریف نوع داده زمان همراه با اختلاف زمانی معادل java.time.OffsetTime
     */
    String COLUMN_DEFINITION_TIME_WITH_TIMEZONE = "TIME_WITH_TIMEZONE";
    /**
     * تعریف نوع داده تاریخ-زمان همراه با اختلاف زمانی معادل java.time.OffsetDateTime
     */
    String COLUMN_DEFINITION_TIMESTAMP_WITH_TIMEZONE = "TIMESTAMP_WITH_TIMEZONE";
    /**
     * تعریف نوع داده کلید اصلی انتیتی در دیتابیس
     */
    String COLUMN_DEFINITION_PRIMARY_KEY = "NUMBER";

    /**
     * این متد برای تبدیل رشته تاریخ به عبارت متناظر آن در دیتابیس اوراکل میباشد
     *
     * @param dateString رشته تاریخ
     * @return خروجی: رشته متناسب تاریخ در دیتابیس اوراکل
     */
    @NotNull
    static String toDate(@NotNull String dateString) {
        if (ObjectUtils.isEmpty(dateString)) {
            dateString = "null";
        }
        return "TO_DATE('" + dateString + "','YYYY-MM-DD HH24:MI:SS')";
    }

    /**
     * این متد برای تبدیل Date تاریخ به عبارت متناظر آن در دیتابیس اوراکل میباشد
     *
     * @param date Date تاریخ
     * @return خروجی: رشته متناسب تاریخ در دیتابیس اوراکل
     */
    @NotNull
    static String toDate(@NotNull Date date) {
        if (!ObjectUtils.isEmpty(date)) {
            return toDate(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(date));
        } else {
            return toDate("null");
        }
    }

    /**
     * این متد ، متد fixQueryIN را با شرط IN فراخوانی میکند
     *
     * @param jpqlFieldName نام فیلد مورد نظر برای کوئری
     * @param itemCsv       مقادیر csv
     * @return خروجی: کوئری مناسب که مشکل محدودیت تعداد csv های دستور IN در اوراکل را نداشته باشد
     */
    static String fixQueryIN(String jpqlFieldName, String itemCsv) {
        return fixQueryIN(jpqlFieldName, itemCsv, " IN ");
    }

    /**
     * این متد ، متد fixQueryIN را با شرط NOT IN فراخوانی میکند
     *
     * @param jpqlFieldName نام فیلد مورد نظر برای کوئری
     * @param itemCsv       مقادیر csv
     * @return خروجی: کوئری مناسب که مشکل محدودیت تعداد csv های دستور IN در اوراکل را نداشته باشد
     */
    static String fixQueryNOTIN(String jpqlFieldName, String itemCsv) {
        return fixQueryIN(jpqlFieldName, itemCsv, " NOT IN ");
    }

    /**
     * این متد به دلیل محدودیتی که اوراکل در استفاده از دستور IN دارد نام فیلد و تعداد مقادیر به صورت csv و نوع IN یا NOTIN را میگیرد و در صورتی که تعداد مقادیر csv بیش از تعداد مجاز اوراکل باشد به چند شرط تبدیل میکند
     *
     * @param jpqlFieldName نام فیلد مورد نظر برای کوئری
     * @param itemCsv       مقادیر csv
     * @param inOrNotIn     کلمه IN یا NOTIN
     * @return خروجی: کوئری مناسب که مشکل محدودیت تعداد csv های دستور IN در اوراکل را نداشته باشد
     */
    static String fixQueryIN(String jpqlFieldName, String itemCsv, String inOrNotIn) {
        List<String> itemList = new ArrayList<>(Arrays.asList(itemCsv.split(",")));
        int i = 1;
        Integer counter = 1;
        StringBuilder resultInStringBuilder = new StringBuilder();
        StringBuilder resultStringBuilder = new StringBuilder();
        for (String item : itemList) {
            //تولید csv مقادیر
            if (!resultInStringBuilder.toString().equals("")) {
                resultInStringBuilder.append(",");
            }
            resultInStringBuilder.append(item);
            if (i < 1000) {
                //تا زمانی که تعداد مقادیر csv به 1000 نرسیده به i اضافه میکنیم
                i++;
                //اگر به انتهای لیست رسیدیم به result مقادیر csv آماده شده را اضافه میکنیم
                if (counter.equals(itemList.size())) {
                    if (!resultStringBuilder.toString().equals("")) {
                        resultStringBuilder.append(" OR ");
                    }
                    resultStringBuilder.append("(").append(jpqlFieldName).append(inOrNotIn).append("(").append(resultInStringBuilder.toString()).append("))");
                }
            } else {
                //وقتی تعداد مقادیر csv شده 1000 میشود در else میفتد به result مقادیر csv آماده شده را اضافه میکنیم
                if (!resultStringBuilder.toString().equals("")) {
                    resultStringBuilder.append(" OR ");
                }
                resultStringBuilder.append("(").append(jpqlFieldName).append(inOrNotIn).append("(").append(resultInStringBuilder.toString()).append("))");
                //مقادیر i و resultIN بعد از هر 1000 تا ریست میشوند
                resultInStringBuilder.setLength(0);
                i = 1;
            }
            counter++;
        }
        return "(" + resultStringBuilder.toString() + ")";
    }
}